[iOS] Maps occasionally crashes under auto layout code when presenting a date picker

whsieh opened the pull request:
#### 5187b76b861c8ce52672b4c995045678795ee15b
<pre>
[iOS] Maps occasionally crashes under auto layout code when presenting a date picker
<a href="https://bugs.webkit.org/show_bug.cgi?id=265906">https://bugs.webkit.org/show_bug.cgi?id=265906</a>
<a href="https://rdar.apple.com/119060260">rdar://119060260</a>

Reviewed by Aditya Keerthi.

Add some extra hardening to the logic in `-_scaleDownToFitHeightIfNeeded`, to avoid ever using 0 or
a non-finite value for `targetScale` when adjusting the date picker&apos;s layout constraints and
transform to fit within the popover container view.

* LayoutTests/fast/forms/ios/show-date-input-in-empty-window-expected.txt: Added.
* LayoutTests/fast/forms/ios/show-date-input-in-empty-window.html: Added.
* LayoutTests/resources/ui-helper.js:

Add a layout test to exercise the change, as well as test infrastructure to resize the web view
along with the containing test runner window.

(window.UIHelper.resizeWindowTo):
* Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
(-[WKDatePickerPopoverController _scaleDownToFitHeightIfNeeded]):
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::resizeWindowTo):
* Tools/WebKitTestRunner/UIScriptControllerCommon.cpp:
(WTR::UIScriptControllerCommon::resizeWindowTo):
* Tools/WebKitTestRunner/UIScriptControllerCommon.h:

Canonical link: <a href="https://commits.webkit.org/271610@main">https://commits.webkit.org/271610@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e0e2551fb03de7cc23a19d8f211a564d0ef1c97

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28945 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30300 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31559 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26377 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29518 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9719 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4933 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26402 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29217 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6271 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24841 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5467 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5602 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25866 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32897 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26463 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26298 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31843 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5572 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3743 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29622 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7214 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/25660 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6923 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6057 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6093 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
Looks reasonable. I wonder how we end up receiving zero width view bounds.

aprotyas reviewed (commented):
LGTM! The test coverage is reassuring, though I'd expect a pathological case like that to not happen in the wild? Anyways, this is good.

whsieh committed:
[iOS] Maps occasionally crashes under auto layout code when presenting a date picker
https://bugs.webkit.org/show_bug.cgi?id=265906
rdar://119060260

Reviewed by Aditya Keerthi.

Add some extra hardening to the logic in `-_scaleDownToFitHeightIfNeeded`, to avoid ever using 0 or
a non-finite value for `targetScale` when adjusting the date picker's layout constraints and
transform to fit within the popover container view.

* LayoutTests/fast/forms/ios/show-date-input-in-empty-window-expected.txt: Added.
* LayoutTests/fast/forms/ios/show-date-input-in-empty-window.html: Added.
* LayoutTests/resources/ui-helper.js:

Add a layout test to exercise the change, as well as test infrastructure to resize the web view
along with the containing test runner window.

(window.UIHelper.resizeWindowTo):
* Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
(-[WKDatePickerPopoverController _scaleDownToFitHeightIfNeeded]):
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::resizeWindowTo):
* Tools/WebKitTestRunner/UIScriptControllerCommon.cpp:
(WTR::UIScriptControllerCommon::resizeWindowTo):
* Tools/WebKitTestRunner/UIScriptControllerCommon.h:

Canonical link: https://commits.webkit.org/271610@main
